commands/topchannels.py

```diff
@@ -97,16 +107,54 @@ def calc_channel_score(channel: Channel):
     """simple calculation of a channels value
     the higher the number of members and the less the number of seconds since the last post the higher the channels score
     """
+    since_latest = seconds_since_last_post(channel)
+    if not since_latest:
+        return 0
     num_members = channel.num_members
-    time_delta_in_hours = seconds_since_last_post(channel) / 3600
+    time_delta_in_hours = since_latest / 3600
 
     return num_members * (exp(-time_delta_in_hours))
 
 
-def seconds_since_last_post(channel: Channel) -> float:
-    """return the fraction of days since the last post in a channel
+def seconds_since_last_post(channel: Channel) -> Optional[float]:
+    """return the fraction of days since the last post in a channel, or None if
+    all messages are of filtered/ignored subtypes
     """
-    return (dt.now() - dt.fromtimestamp(channel.latest_ts)).total_seconds()
+    ignore_message_types = {"channel_join"}
+
+    if channel.latest_subtype not in ignore_message_types:
+        latest_ts = channel.latest_ts
+    else:
+        # The latest message in this channel has an ignored subtype, so we need
+        # to dig through history to find the most recent non-ignored message.
+        # Similar to invite permissions, this requires a user token.
+        #
+        # "New" bot tokens will be both more granular and more flexible, so should
+        # be able to replace the need for user tokens going forward:
+        #
+        # Ref: https://api.slack.com/docs/token-types#bot_new
+
+        grant_user_token = os.environ.get("SLACK_KARMA_INVITE_USER_TOKEN")
+        karmabot_id = os.environ.get("SLACK_KARMA_BOTUSER")
+        if not grant_user_token:
+            logging.info(
+                "Cannot search channel history, no env SLACK_KARMA_INVITE_USER_TOKEN"
+            )
+            return None
+
+        sc = SlackClient(grant_user_token)
+        response = sc.api_call("channels.history", channel=channel.id, user=karmabot_id)
+
+        msgs = [
```

Comment:
you could return msgs, if it's empty it's "falsy"

Reply:
Hmm I'm torn on that. The method returns a float. Making it `Optional` to allow for a `None` return seems logical enough, but having it return `float` or a falsy empty list feels awkward.

Maybe it's worth a refactor so we don't call `seconds_since_last_post` at all unless we have a valid (not ignored) message timestamp.

Reply:
Good point, thanks